sdks/python/apache_beam/ml/inference/tensorrt_inference.py

```diff
@@ -0,0 +1,213 @@
+    self.max_batch_size = max_batch_size
+    if 'engine_path' in kwargs:
+      self.engine_path = kwargs.get('engine_path')
+    elif 'onnx_path' in kwargs:
+      self.onnx_path = kwargs.get('onnx_path')
+
+    trt.init_libnvinfer_plugins(LOGGER, namespace="")
+
+  def batch_elements_kwargs(self):
+    """Sets min_batch_size and max_batch_size of a TensorRT engine."""
+    return {
+        'min_batch_size': self.min_batch_size,
+        'max_batch_size': self.max_batch_size
+    }
+
+  def load_model(self) -> trt.ICudaEngine:
+    """Loads and initializes a TensorRT engine for processing."""
+    return _load_engine(self.engine_path)
+
+  def load_onnx(self) -> trt.INetworkDefinition:
+    """Loads and parses an onnx model for processing."""
+    return _load_onnx(self.onnx_path)
+
+  #
+  def build_engine(self, network: trt.INetworkDefinition) -> trt.ICudaEngine:
+    """Build an engine according to parsed/created network."""
+    return _build_engine(network)
+
+  def run_inference(
+      self,
+      batch: np.ndarray,
+      engine: trt.ICudaEngine,
+      inference_args: Optional[Dict[str, Any]] = None
+  ) -> Iterable[PredictionResult]:
+    """
+    Runs inferences on a batch of Tensors and returns an Iterable of
+    TensorRT Predictions.
+
+    Args:
+      batch: A np.ndarray or a np.ndarray that represents a concatenation
+        of multiple arrays as a batch.
+      engine: A TensorRT engine.
+      inference_args: Any additional arguments for an inference
+        that are not applicable to TensorRT.
+
+    Returns:
+      An Iterable of type PredictionResult.
+    """
+    _validate_inference_args(inference_args)
+    context = engine.create_execution_context()
+    assert context
+    # Setup I/O bindings
+    inputs = []
+    outputs = []
+    allocations = []
+    for i in range(engine.num_bindings):
+      is_input = False
+      if engine.binding_is_input(i):
+        is_input = True
+      name = engine.get_binding_name(i)
+      dtype = engine.get_binding_dtype(i)
+      shape = engine.get_binding_shape(i)
+      if is_input:
+        batch_size = shape[0]
+      size = np.dtype(trt.nptype(dtype)).itemsize
+      for s in shape:
+        size *= s
+      allocation = cuda.mem_alloc(size)
+      binding = {
+          'index': i,
+          'name': name,
+          'dtype': np.dtype(trt.nptype(dtype)),
+          'shape': list(shape),
+          'allocation': allocation,
+      }
+      allocations.append(allocation)
+      if engine.binding_is_input(i):
+        inputs.append(binding)
+      else:
+        outputs.append(binding)
+
+    assert batch_size > 0
+    assert len(inputs) > 0
+    assert len(outputs) > 0
+    assert len(allocations) > 0
+    # Prepare the output data
+    predictions = []
+    for output in outputs:
+      predictions.append(np.zeros(output['shape'], output['dtype']))
+    # Process I/O and execute the network
+    cuda.memcpy_htod(inputs[0]['allocation'], np.ascontiguousarray(batch))
```

Comment:
Is this only intended to work with single-input models?

Reply:
It would be pretty much the same way you handle the outputs. Instead of passing in a single input, you could take a `feed_dict` argument (e.g. a `Dict[str, np.ndarray]`). Then when you memcpy, you would have something like:
```py
for name, buffer in feed_dict.items():
    input_buffer = ... # Figure out which input buffer to copy to using, e.g. `get_binding_name()`
    cuda.memcpy_htod(input_buffer, np.ascontiguousarray(buffer))
```

You can also refer to [how Polygraphy does it](https://github.com/NVIDIA/TensorRT/blob/main/tools/Polygraphy/polygraphy/backend/trt/runner.py#L221) for a more comprehensive example.

Reply:
@pranavm-nvidia ok, definitely need to come up with a use case to test it out. Thank you!

Reply:
I'll keep this in mind, will be added in future releases.